File: backend/src/main/java/tom/projects/monkeylog/controller/ExerciseGroupController.java
package tom.projects.monkeylog.controller;

import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import tom.projects.monkeylog.dto.workout.*;
import tom.projects.monkeylog.mapper.WorkoutMapper;
import tom.projects.monkeylog.service.ExerciseGroupService;
import tom.projects.monkeylog.service.ExerciseRowService;

import java.util.UUID;

@RestController
@RequiredArgsConstructor
@Tag(name = "Workout")
public class ExerciseGroupController {
    private final ExerciseGroupService exerciseGroupService;
    private final ExerciseRowService exerciseRowService;
    private final WorkoutMapper workoutMapper;

    @PostMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}/exercise-rows")
    ExerciseGroupResponse createExerciseRow(@PathVariable UUID workoutId,
                                            @PathVariable UUID exerciseGroupId) {
        return workoutMapper.exerciseGroupToExerciseGroupResponse(exerciseGroupService.addRow(exerciseGroupId));
    }

    @PatchMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}/exercise-rows/{exerciseRowId}")
    ExerciseRowResponse updateExerciseRow(@PathVariable UUID workoutId,
                                          @PathVariable UUID exerciseGroupId,
                                          @PathVariable UUID exerciseRowId,
                                          @RequestBody ExerciseRowUpdateRequest exerciseRowUpdateRequest) {
        return workoutMapper.exerciseRowToExerciseRowResponse(exerciseRowService.update(exerciseRowUpdateRequest, exerciseRowId));
    }

    @PatchMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}/exercise-rows/{exerciseRowId}/exercise-row-fields/{exerciseRowFieldId}")
    ExerciseRowFieldResponse updateExerciseRowField(@PathVariable UUID workoutId,
                                                    @PathVariable UUID exerciseGroupId,
                                                    @PathVariable UUID exerciseRowId,
                                                    @PathVariable UUID exerciseRowFieldId,
                                                    @RequestBody ExerciseRowFieldUpdateRequest exerciseRowFieldUpdateRequest) {
        return workoutMapper.exerciseRowFieldToExerciseRowFieldResponse(
                exerciseRowService.updateField(exerciseRowFieldUpdateRequest, exerciseRowFieldId));
    }

    @PostMapping("/workouts/{workoutId}/exercise-groups")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    void createExerciseGroup(@PathVariable UUID workoutId, @RequestBody ExerciseGroupCreateRequest exerciseGroupCreateRequest) {
        exerciseGroupService.save(exerciseGroupCreateRequest, workoutId);
    }

    @DeleteMapping("/workouts/{workoutId}/exercise-group/{exerciseGroupId}/exercise-rows/{exerciseRowId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    void deleteExerciseRow(@PathVariable UUID workoutId, @PathVariable UUID exerciseGroupId, @PathVariable UUID exerciseRowId) {
        exerciseRowService.delete(exerciseRowId);
    }

    @DeleteMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    void deleteExerciseGroup(@PathVariable UUID workoutId, @PathVariable UUID exerciseGroupId) {
        exerciseGroupService.delete(exerciseGroupId);
    }
}

File: backend/src/main/kotlin/tom/com/monkeylog/controller/ExerciseGroupController.kt
package tom.com.monkeylog.controller

import io.swagger.v3.oas.annotations.tags.Tag
import org.springdoc.core.annotations.ParameterObject
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import tom.com.monkeylog.common.dto.IdProjection
import tom.com.monkeylog.common.dto.Page
import tom.com.monkeylog.common.mapper.toResponse
import tom.com.monkeylog.dto.workout.*
import tom.com.monkeylog.mapper.toResponse
import tom.com.monkeylog.mapper.toResponseWithWorkout
import tom.com.monkeylog.service.ExerciseGroupService
import tom.com.monkeylog.service.ExerciseRowService
import java.util.*

@RestController
@Tag(name = "Workout")
class ExerciseGroupController(
    private val exerciseGroupService: ExerciseGroupService,
    private val exerciseRowService: ExerciseRowService,
) {
    @PostMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}/exercise-rows")
    fun createExerciseRow(
        @PathVariable workoutId: UUID,
        @PathVariable exerciseGroupId: UUID
    ) = exerciseGroupService.addRow(exerciseGroupId).toResponse()

    @PutMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}/exercise-rows/{exerciseRowId}")
    fun updateExerciseRow(
        @PathVariable workoutId: UUID,
        @PathVariable exerciseGroupId: UUID,
        @PathVariable exerciseRowId: UUID,
        @RequestBody exerciseRowUpdateRequest: ExerciseRowUpdateRequest
    ) = exerciseRowService.update(
        exerciseRowUpdateRequest,
        exerciseRowId
    ).toResponse()

    @PostMapping("/workouts/{workoutId}/exercise-groups")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun createExerciseGroup(
        @PathVariable workoutId: UUID,
        @RequestBody exerciseGroupCreateRequest: ExerciseGroupCreateRequest
    ) = exerciseGroupService.save(exerciseGroupCreateRequest, workoutId)

    @DeleteMapping("/workouts/{workoutId}/exercise-group/{exerciseGroupId}/exercise-rows/{exerciseRowId}")
    @ResponseStatus(
        HttpStatus.NO_CONTENT
    )
    fun deleteExerciseRow(
        @PathVariable workoutId: UUID,
        @PathVariable exerciseGroupId: UUID,
        @PathVariable exerciseRowId: UUID
    ) = exerciseRowService.delete(exerciseRowId)

    @DeleteMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteExerciseGroup(@PathVariable workoutId: UUID, @PathVariable exerciseGroupId: UUID) =
        exerciseGroupService.delete(exerciseGroupId)

    @GetMapping("/exercises/{exerciseId}/exercise-groups")
    fun getExerciseGroups(
        @PathVariable exerciseId: UUID,
        @ParameterObject pageable: Pageable
    ): Page<ExerciseGroupWithWorkoutResponse> {
        val uuids = exerciseGroupService.getExerciseGroupsByExerciseId(exerciseId, pageable)
        val exerciseGroups =
            exerciseGroupService.getExerciseGroups(uuids.content.map(IdProjection::id))
                .associate { it.id!! to it.toResponseWithWorkout() }

        return uuids.toResponse { exerciseGroups[it.id]!! }
    }
}

